mdbook/src/chapter_4/chapter_4_2.md

```diff
@@ -2,16 +2,17 @@
 
 As an overview, timely dataflow's progress tracking is about (i) maintaining a view of *timestamp capabilities* at each location in the dataflow graph, and (ii) communicating the implications of changes in capabilities to other locations in the dataflow graph. Before we get in to these two aspects, we will first need to be able to name parts of our dataflow graph.
 
 ## Dataflow structure
 
-A dataflow graph hosts some number of operators. For progress tracking, these operators are simply identified by their index. Each operator has some number of *input ports*, each of which may independently have or not have certain timestamps in its future. Each operator has some number of *output ports*, each of which may independently express a capability to produce timestamps in the future. Each input port is connected to a single output port, and each output port may be connected to multiple input ports (a message produced at an output port is to be delivered to all attached input ports).
+A dataflow graph hosts some number of operators. For progress tracking, these operators are simply identified by their index. Each operator has some number of *input ports*, each of which may independently have or not have certain timestamps in its future [not clear what "have timestamps" and "in its future" means]. Each operator has some number of *output ports*, each of which may independently express a capability to produce [data with] timestamps in the future [again, not clear what "future" is]. Each input port is connected to a single output port, and each output port may be connected to multiple input ports (a message produced at an output port is to be delivered to all attached input ports)[should make clear that you mean the output hooking up to someone else's input. As written it sounds cyclic].
 
-In timely dataflow progress tracking, we identify output ports by the type `Source` and input ports by the type `Target`, as an output port is a source of timestamped data, and an input port is a target of timestamped data. Each can be described by their operator index and then an operator-local index of the corresponding port. The use of distinct types helps us avoid mistaking input and output ports.
+In timely dataflow progress tracking, we identify output ports by the type `Source` and input ports by the type `Target`, as an output port is a source of timestamped data, and an input port is a target of timestamped data[this nomenclature is confusing, but hopefully less so in the actual code? Would be weird if operators used their "source" to write to]. Each can be described by their operator index and then an operator-local index of the corresponding port. The use of distinct types helps us avoid mistaking input and output ports.
 
+[hmm, I would really think that data comes from Source and goes to Target. Is it really the other way?]
```

Comment:
It is, because we are talking about all of this from the perspective of the manager of the dataflow graph. It doesn't know how the operators' internals work, just that the outputs of operators produce data, and the input consume data. Unmentioned in the text, there are other sources and targets too (e.g. in a nested scope, the "ingress" nodes are sources, and the "egress" nodes are targets. They look like operator outputs and inputs, respectively.